hathor/debug_resources.py

```diff
@@ -0,0 +1,223 @@
+
+@register_resource
+class DebugLogResource(resource.Resource):
+    isLeaf = True
+    openapi = {
+        '/_debug/log': {
+            'x-visibility': 'private',
+            'get': {
+                'operationId': 'debug_log',
+                'summary': 'Used for forcing any log message',
+            }
+        }
+    }
+    valid_log_levels = {
+        'critical',
+        'debug',
+        'error',
+        'exception',
+        'fatal',
+        'info',
+        'msg',
+        'warn',
+        'warning',
+    }
+    default_log_level = 'info'
+    default_log_msg = 'debugging logging system'
+
+    def render_GET(self, request):
+        logger_name = get_arg_default(request.args, 'logger', None)
+        if logger_name is not None:
+            log = get_logger(logger_name).new()
+        else:
+            log = logger.new()
+        level = get_arg_default(request.args, 'level', self.default_log_level)
+        assert level in self.valid_log_levels
+        log_func = getattr(log, level)
+        msg = get_arg_default(request.args, 'msg', self.default_log_msg)
+        # TODO: maybe add an `extras` param (probably as a json body via POST) to add arbitray structed attributes
+        log_func(msg)
+        return b'OK'
+
+
+@register_resource
+class DebugMessAroundResource(resource.Resource):
+    isLeaf = True
+    openapi = {
+        '/_debug/mess_around': {
+            'x-visibility': 'private',
+            'get': {
+                'operationId': 'debug_mess_around',
+                'summary': 'Used for forcing an error on the full-node by messing with stuff',
+            }
+        }
+    }
+    default_mess = 'storage'
+
+    def __init__(self, manager: HathorManager):
+        super().__init__()
+        self.manager = manager
+        # map is create here to be able to access the bound methods
+        self.mess_map = {
+            'storage': self.remove_storage,
+        }
+
+    def remove_storage(self):
+        # XXX: this is just used to cause a problem on another part of the fullnode
+        self.manager.tx_storage = None
+
+    def render_GET(self, request):
+        mess = get_arg_default(request.args, 'with', self.default_mess)
+        assert mess in self.mess_map
+        mess_func = self.mess_map[mess]
+        reactor.callFromThread(mess_func)
+        return b'OK'
+
+
+@register_resource
+class DebugCrashResource(resource.Resource):
+    isLeaf = True
+    openapi = {
+        '/_debug/crash': {
+            'x-visibility': 'private',
+            'get': {
+                'operationId': 'debug_crash',
+                'summary': 'Used for forcing an unclean exit, this will almost certainly corrupt the database',
+            }
+        }
+    }
+
+    def run(self, code: int) -> None:
+        # XXX: sys.exit will raise a SystemExit exception that get's trapped by twisted
+        #      os._exit will bypass that by exiting directly, note that no cleanup methods will be called
+        os._exit(code)
```

Comment:
Isn't it the same as using `kill -9`?

Reply:
The main difference is that `kill -9` isn't able to make the process seem it exited successfully. With `os._exit` we can exit with `0` and it will seem like a "successful" exit. I have no idea if this is actually useful, but might be (a successful exit may signal that the database was closed successfully).

Reply:
Ok. Thanks!